### PR TITLE
[ruff] Update 0.0.255 -> 0.0.277

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Internal companion PR: https://github.com/dagster-io/internal/pull/5846

Updates ruff from 0.0.255 to 0.0.277.

This introduces autofix for the TCH rule, which allows auto-sorting type-checking only imports into `TYPE_CHECKING` blocks. Because the changes resulting from enabling this rule are large, they are in an upstack PR.

This PR contains only the ruff version update, a few fixes for issues surfaced by ruff behavior changes to already-enabled rules, and the new rules RUF009 and RUF010 (which are enabled because we enable the whole category of RUF* rules).

- RUF009 flags function calls in field defaults for dataclasses
- RUF010 converts e.g. this: `f"foo {repr(some_var)}"` to the more idiomatic `f"foo {some_var!r}"` using [f-string conversion flags](https://peps.python.org/pep-0498/#specification).

## How I Tested These Changes

ruff

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #14674** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14674" target="_blank">Graphite</a>  👈
    * **PR #14675** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/14675" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/14674?utm_source=stack-comment).

File: python_modules/dagster/dagster_tests/storage_tests/test_local_instance.py [hunk 1]
```diff
@@ -284,16 +286,31 @@ def _instantiate_ephemeral_instance(_):
...
+
+        def _run(_):
+            shared.root_directory  # noqa: B018
+            with DagsterInstance.ephemeral() as instance:
+                instance.root_directory  # noqa: B018
+            return True
+
+        with ThreadPoolExecutor(max_workers=n) as executor:
+            results = executor.map(_run, range(n))
+            assert all(results)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster_tests/storage_tests/test_local_instance.py:301]: ? bad rebase? don't add this back

File: python_modules/libraries/dagster-azure/dagster_azure/adls2/__init__.py [hunk 2]
```diff
@@ -14,7 +14,6 @@
 )
 from .resources import (
     ADLS2DefaultAzureCredential as ADLS2DefaultAzureCredential,
-    ADLS2FileManager as ADLS2FileManager,
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-azure/dagster_azure/adls2/__init__.py:17]: ?

**Review by alangenfeld — APPROVED** [review]

File: python_modules/dagster/dagster_tests/storage_tests/test_local_instance.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster_tests/storage_tests/test_local_instance.py:301]: yeah must've been a mistake in rebase conflict resolution

File: python_modules/libraries/dagster-azure/dagster_azure/adls2/__init__.py (hunk 2 above)
Comment (smackesey) [reply python_modules/libraries/dagster-azure/dagster_azure/adls2/__init__.py:17]: This is deleted because `ADLS2FileManager` is already exported above.

**Pull request merged**